include/l3.h

```diff
@@ -55,14 +55,57 @@ int l3_init(const char *path);
 
 /**
  * \brief Caller-macro to invoke L3 simple logging.
+ *
+ * Macro to synthesize LOC-encoding value, if needed, under conditional
+ * compilation flags. DEBUG-version of macros are provided to cross-check
+ * that the supplied format string in `msg` are consistent with the
+ * arguments' type.
  */
-#ifdef L3_LOC_ENABLED
-#define l3_log_simple(msg, arg1, arg2)                          \
-        l3__log_simple(__LOC__, (msg), (arg1), (arg2))
-#else   // L3_LOC_ENABLED
-#define l3_log_simple(msg, arg1, arg2)                          \
-        l3__log_simple(L3_ARG_UNUSED, (msg), (arg1), (arg2))
-#endif  //
+#if defined(DEBUG)
```

Comment:
This is a pattern you will see in this change-set.

Redefine caller-macros under `DEBUG` build mode, so that we eventually compile `printf()`, which will trip-up if calling code is wrong.

Reply:
You can use `__attribute__((format(printf, x, y)))` which will invoke the compiler warnings. Might be a bit simpler and potentially give better compiler errors? Supported by both gcc and clang.

Reply:
Thanks for this tip, @gregthelaw -- I learnt something new today!

I tried this out ... but found out thru experimentation, and after reading [the docs carefully](https://gcc.gnu.org/onlinedocs/gcc-3.2/gcc/Function-Attributes.html), that this syntax is valid only if the argument lists are defined as varargs, using `...`.

See this doc note:


format (archetype, string-index, first-to-check)
    The format attribute specifies that a function takes printf, scanf, strftime or strfmon style arguments which should be type-checked against a format string. For example, the declaration:

              extern int
              my_printf (void *my_object, const char *my_format, ...)
                    __attribute__ ((format (printf, 2, 3)));
              

    causes the compiler to check the arguments in calls to my_printf for consistency with the printf style format string argument my_format. 
    
    ---

NOTE that the example shows: `const char *my_format, ...)` ... which means var-args are being used.

----

When I tried to implement this attribute, I get this compilation error:

```
gcc  -x c -DL3_ENABLED -D_GNU_SOURCE -ggdb3 -Wall -Wfatal-errors -Werror -I ./include -I ./tests/unit/ -c tests/unit/l3_dump.py-test.c -o build/release/obj/tests/unit/l3_dump.py-test.o
In file included from tests/unit/l3_dump.py-test.c:27:
./include/l3.h:131:21: error: ‘format’ attribute argument 3 value ‘2’ does not refer to a variable argument list
  131 |                     const uint32_t loc) __attribute__ ((format(__printf__, 1, 2)));
      |                     ^~~~~
compilation terminated due to -Wfatal-errors.
```

----

I resisted converting the log-msg interface to use var-args right now, in this change-set.

That extension is something I have in mind, in a future change, once we get the rest of core perf-benchmarking working.

Reply:
Ahh, makes sense!